### ISSUE TITLE
About 360VR video output

### ISSUE THREAD

**Issue by T-Photoworks** [NVlabs/instant-ngp#979]
hello　everyone
A few questions have been asked here but there doesn't seem to be an answer. But I think there must be a way, so I ask again.
If the lens distortion on the Instant-ngp screen is set to LatLong, the training screen becomes a 1:2 LatLong screen.
But when I output it as a LatLong video from the camera path, it's not 1:2 LatLong , it's cropped a bit and it's not a 360VR video.

I've found some code, but I'm not familiar with Python code or anything else to know how to do it.
for example.
--datadir <myscene>-dataset_type llffconfig_fern.txt --no_ndc --spherify --lindisp
testbed.nerf.render_distortion.mode = ngp.CameraDistortionMode.LatLong
etc., 
Is this code what I want? If so, I don't know what to do with it.
Thank you.

**Comment by AlbertoOddness** [comment]
Hi!

I've produced 360 videos so I'll explain you what I did. There might be better ways but this was sufficiently simple. 

First, go to instant-ngp/scripts/run.py and in line 146 add this
```
	testbed.nerf.render_distortion.mode = ngp.LensMode.LatLong
```
As a reference, the surrounding lines of code should look something like
```
	testbed.shall_train = args.train if args.gui else True


	testbed.nerf.render_with_lens_distortion = True
	testbed.nerf.render_distortion.mode = ngp.LensMode.LatLong
    
	network_stem = os.path.splitext(os.path.basename(network))[0]
	if args.mode == "sdf":
		setup_colored_sdf(testbed, args.scene)
```

This way the video will be rendered as LatLong. To get the 2:1 ratio, when you run the script to render you need to specify width and height that are in 2:1. Here's an example line of how you would do it from the terminal with the fox sample data once you generate a snapshot and a camera path:

```
instant-ngp$ python scripts/run.py --mode nerf --scene data/nerf/fox --load_snapshot data/nerf/fox/base.msgpack --video_camera_path data/nerf/fox/base_cam.json --video_n_seconds 10 --video_fps 30 --width 1000 --height 500
```

Width and height can be any size you like as long as they're 2:1. Now, this output lacks the correct metadata for video players like VLC or YouTube to recognize it as a 360 video, so they'll just display it as a flat video.

There's different ways to inject the metadata into the resulting video. The way I did it was using [this repository](https://github.com/google/spatial-media). Clone it, move to the spatialmedia subdirectory and follow their instructions on how to run it. If you're using linux, try follow the instructions [provided in this post](https://github.com/google/spatial-media/issues/248#issuecomment-927349482). This one was easy enough so I didn't try other methods but googling "inject spherical metadata to 360 video" will bring up plenty of other options.

Good luck! Any further questions and I'll be happy to try to help 🙂️

**Comment by T-Photoworks** [comment]
Hi,
I'm glad to hear that LatLong output can be done easily.Thank you.
However, when I tried it, I got an error message and could not render. ”File "E:\instant-ngp\instant-ngp\scripts\run.py”, line 147testbed.nerf.render_distortion.mode = ngp.LensMode.LatLong"
is displayed.
Also, ”testbed.nerf.render_with_lens_distortion = True” on line 145
Where is with_camera_distortion in my case, is this old code?

**Comment by AlbertoOddness** [comment]
> Hi, I'm glad to hear that LatLong output can be done easily.Thank you. However, when I tried it, I got an error message and could not render. ”File "E:\instant-ngp\instant-ngp\scripts\run.py”, line 147testbed.nerf.render_distortion.mode = ngp.LensMode.LatLong" is displayed. Also, ”testbed.nerf.render_with_lens_distortion = True” on line 145 Where is with_camera_distortion in my case, is this old code?

Yes, it seems like you're running an older version of the project. Try cloning it again now to get the latest version, that should fix this error.

**Comment by T-Photoworks** [comment]
Hi,
When you output LatLong, were you able to output at 180 x 360 in height and width?  In my case, the video is a little lacking in top, bottom, left, and right. Thank you.
Tomonori Taniguchi 
iPadから送信

**Comment by AlbertoOddness** [comment]
> Hi, When you output LatLong, were you able to output at 180 x 360 in height and width? In my case, the video is a little lacking in top, bottom, left, and right. Thank you. Tomonori Taniguchi iPadから送信


With those specific dimensions? I've not tried, but I can give it a go. I think it should work just fine though, maybe since the dimensions are very small it'll look too distorted. I've tried making videos at 1200x600 or other similar dimensions and it worked fine, with no issues.
